hudi-common/src/main/java/org/apache/hudi/common/table/read/FileGroupRecordBuffer.java

```diff
@@ -498,32 +482,32 @@ protected Option<T> merge(Option<T> older, Map<String, Object> olderInfoMap,
               } else {
                 indexedRecord = (IndexedRecord) mergedRecord.get().getLeft().getData();
               }
-              return Option.ofNullable(readerContext.convertAvroRecord(indexedRecord));
+              return Pair.of(false, readerContext.convertAvroRecord(indexedRecord));
             }
-            return Option.empty();
+            return null;
           } else {
-            if (older.isEmpty() || newer.isEmpty()) {
-              if (shouldKeepNewerRecord(older, olderInfoMap, newer, newerInfoMap)) {
+            if (olderRecord.isDelete() || newerRecord.isDelete()) {
+              if (shouldKeepNewerRecord(olderRecord, newerRecord)) {
                 // IMPORTANT:
                 // this is needed when the fallback HoodieAvroRecordMerger got used, the merger would
                 // return Option.empty when the new payload data is empty(a delete) and ignores its ordering value directly.
-                return newer;
+                return Pair.of(newerRecord.isDelete(), newerRecord.getRecord());
               } else {
-                return older;
+                return Pair.of(olderRecord.isDelete(), olderRecord.getRecord());
               }
             }
             Option<Pair<HoodieRecord, Schema>> mergedRecord = recordMerger.get().merge(
-                readerContext.constructHoodieRecord(older, olderInfoMap), readerContext.getSchemaFromMetadata(olderInfoMap),
-                readerContext.constructHoodieRecord(newer, newerInfoMap), readerContext.getSchemaFromMetadata(newerInfoMap), props);
+                readerContext.constructHoodieRecord(olderRecord), readerContext.getSchemaFromBufferRecord(olderRecord),
```

Comment:
One thing to validate later is that custom merger implementation can return `HoodieEmptyRecord` to indicate deletes.  That should be properly handled in the CUSTOM merge mode.